### ISSUE TITLE
Issue with $wgLinkTitlesSkipTemplates = true;

### ISSUE THREAD

**Issue by Ikabodo** [bovender/LinkTitles#29]
I have added $wgLinkTitlesSkipTemplates = true; to Localsettings.php, and it's mostly working fine. But sometimes it still autolinks things within tags.

Here is an example of code where "Hillfort" is getting autotagged. After some testing I believe that your code considers the end of "{{PAGENAME}}" (the "}}"-part) to be the end of the parser. This means that any parser with magic words or similar inside of it will still be autotagged. This causes some issues with infobox related templates.

```
{{Infobox_manor
|Image=
|Name={{PAGENAME}}
|Other names=Longbarrows
|Holder=Ren
|Steward=Pryderi
|Hundred=[[Hillfort]]
}}
```

**Comment by bovender** [comment]
At first sight you're probably right. I don't think the regular expression respects nested braces/parentheses.

**Comment by Ikabodo** [comment]
If it's not possible to change this, perhaps it's possible to add a <noautolink> </noautolink> or something similar? That way the user can choose a part of the page not to autolink if issues arrise. Sort of like the magic word but only for a certain part of the page.

**Comment by bovender** [comment]
Oh, I guess it's possible to tweak the regex, it just requires a bit more regex voodoo that I can do off the top of my head. I'll research this.

**Comment by Ikabodo** [comment]
A small addition after some testing. The same happens in `<gallery></gallery>` tags. Perhaps something to add to "do not parse".

**Comment by Ikabodo** [comment]
Since i wrote this issue, I've developed my wiki quite a bit. And realized that there might very well be use for a separate `<autolink></autolink>` and a ` <noautolink></noautolink>`. I'm using a lot of Page Forms, where you do not want to parse some of the results, but some parts you definitely do (text based parts). It would be extra useful if able to be put anywhere, i.e. the `<autolink></autolink>` FORCES the text within them to be parsed. 

It would be a good way for a user to customize where they do and do not want the pages parsed.

**Comment by bovender** [comment]
I'll add support for `<noautolink>...</noautolink>`.

Can you please clarify what happens inside `<gallery>` tags. They should be excluded from automatic linking at this point already, so I wonder what's going wrong.

With regard to `<autolink>...</autolink>` I'm not so sure. If you go out of your way to add these tags, wouldn't it be easier to just go ahead and link the relevant page titles yourself? Also, this would add considerable complexity to the code: Look for the `__NOAUTOLINKS__` magic word and if it exists, look again for any `<autolink>` tags, and if there are any, process only that part of the page.

**Comment by Ikabodo** [comment]
**Regarding `<noautolink>`:**
Cool! Thank you for taking the time to create such an awsome extension. 

**Regarding `<gallery>`:**
I've had examples where the extension autolinks texts inside `<gallery></gallery>` tags, I'm not exactly sure what causes it as I can't seem to replicate it right now. I'll try to replicate it and get back to you soon.

**Regarding `<autolink>`:**
I still think it's a good idea, but you make a good point. This is my scenario:
_I use Page Forms, a very usefull extension of the Semantic Mediawiki project. In this extension you fill out a form that then creates templates based on what you fill out in the form. Some of the values in this form are texts, not just simple words. Consider it a form the user fills out, with some small text boxes, a dropdown and a textarea. You would want the textboxes and the dropdown not to create autolinks, but the textarea you want autolinked._ 

The template below is an example where I would want the |Description= part autolinked, but not the |Mother= and |Father= parts. 
```
{{CharacterSections
|Mother=Unknown
|Father=Nanteleod
|Description text=Sir Alain de Carleon is a jovial knight from the kingdom of Escavalon. He is a loyal vassal of King Nanteleod and holds the city of Carlion for his liege.

Sir Alain is handsome, jovial and a fan of ale and women. Sir Alain lives life to its fullest, never passing an opportunity to drink, fight, or be merry. He is a famed swordsman and an even more famed boaster, but always remembering to include his friends in his tales (whether they were present at the feat or not).
}}
```

But then again, this could be solved with the `<noautolink>` you described, by removing the "do not autolink templates" options, and then manually adding `<noautolink>` to every minor part. But I believe the parts you WOULD want autotaged is far fewer than the other way around. Hence the `<autolink>` parts. 

I hope I managed to explain the scenario.

**Comment by bovender** [comment]
OK, so adding support for `<noautolink>` is not as straightforward as it seemed; I need to find a way to eliminate the tags from the output.

Your point about the use case for `<autolink>` is well taken.

As for the initial issue with template tags that contain nested `{{...}}` tags, I have made some progress (using a recursive regular expression), but need to find a way to allow linking inside template parameter values (for those users who have `$wgLinkTitlesSkipTemplates = false;`).

**Comment by bovender** [comment]
@Ikabodo I've released version 4.1.0 which comes with support for `<autolinks>...</autolinks` and `<noautolinks>...</noautolinks>` tags (cf. [documentation](https://www.mediawiki.org/wiki/Extension:LinkTitles#Enable_or_disable_automatic_linking_for_sections)), and which fixes the issue with the templates.
